farm/data_handler/processor.py

```diff
@@ -1671,6 +1672,168 @@ def _sample_to_features(self, sample) -> dict:
+        :param train_filename: The name of the file containing training data.
+        :type train_filename: str
+        :param dev_filename: The name of the file containing the dev data. If None and 0.0 < dev_split < 1.0 the dev set
+                             will be a slice of the train set.
+        :type dev_filename: str or None
+        :param test_filename: None
+        :type test_filename: str
+        :param dev_split: The proportion of the train set that will sliced. Only works if dev_filename is set to None
+        :type dev_split: float
+        :param proxies: proxy configuration to allow downloads of remote datasets.
+                        Format as in  "requests" library: https://2.python-requests.org//en/latest/user/advanced/#proxies
+        :type proxies: dict
+        :param embed_title: Whether to embed title in passages during tensorization (bool),
+        :param num_hard_negatives: maximum number to hard negative context passages in a sample
+        :param num_positives: maximum number to positive context passages in a sample
+        :param kwargs: placeholder for passing generic parameters
+        :type kwargs: object
+        """
+        #TODO If an arg is misspelt, e.g. metrics, it will be swallowed silently by kwargs
+
+        # Custom processor attributes
+        self.max_samples = max_samples
+        self.query_tokenizer = query_tokenizer
+        self.passage_tokenizer = passage_tokenizer
+        self.embed_title = embed_title
+        self.num_hard_negatives = num_hard_negatives
+        self.num_positives = num_positives
+
+        super(DPRProcessor, self).__init__(
+            tokenizer= query_tokenizer,
+            max_seq_len=max_seq_len,
+            train_filename=train_filename,
+            dev_filename=dev_filename,
+            test_filename=test_filename,
+            dev_split=dev_split,
+            data_dir=data_dir,
+            tasks={},
+            proxies=proxies,
+        )
+        if metric:
+            task_type = "dpr"
+            self.add_task(name="dpr",
+                          metric=metric,
+                          label_list=None,
+                          label_name=None,
+                          task_type=task_type)
+        else:
+            logger.info("Initialized processor without tasks. Supply `metric` and `label_list` to the constructor for "
+                        "using the default task or add a custom task later via processor.add_task()")
+
+    def file_to_dicts(self, file: str) -> [dict]:
+        # TO-DO: json should be in SQUAD format
+        dicts = json.load(open(file))
+        return dicts
+
+    def _normalize_question(self, question: str) -> str:
+        if question[-1] == '?':
+            question = question[:-1]
+        return question
+
+    def _dict_to_samples(self, dictionary: dict, **kwargs) -> [Sample]:
+        """
+        dictionary representing 1 sample to Sample for DPR
+        :param dictionary:  {"question": str,
```

Comment:
It might make sense here to rather structure it like this: 
```
  {
      "query": query_text
       "passages": [{"text": document_text, "title": xxx, "label": "positive", "external_id": abb123}]
}
   ```

These dicts will be also used later at inference time where we don't really have a notion of "positive" and "hard_negative" passages. We will just have queries and passages to embed.  I believe the above structure is therefore more flexible and generic.

Reply:
fixed it in files_to_dict and dict_to_samples